lib/argon2.js

```diff
@@ -116,9 +116,16 @@
             var MB = 1024 * KB;
             var GB = 1024 * MB;
             var WASM_PAGE_SIZE = 64 * 1024;
+            /*
+             * Temporary fix to circumvent the impossibility
+             * to grow the amount of allocated memory dynamically
+             * in between each computation: set a higher minimum
+             * initial memory.
+             */
+            var MIN_INITIAL_MEMORY = 64 * MB / WASM_PAGE_SIZE;
 
             var totalMemory = (2 * GB - 64 * KB) / 1024 / WASM_PAGE_SIZE;
-            var initialMemory = Math.min(Math.max(Math.ceil(params.mem * 1024 / WASM_PAGE_SIZE), 256) + 256, totalMemory);
+            var initialMemory = Math.min(Math.max(Math.ceil(params.mem * 1024 / WASM_PAGE_SIZE), MIN_INITIAL_MEMORY) + 256, totalMemory);
```

Comment:
The old value was 256, now it's 1024 * 1024, is it what you intended to do? Doesn't feel right :thinking:

Reply:
Now it's 64 * 1024 * 1024 / (64 * 1024) => 1024

Reply:
It's the number of WASM pages.

Reply:
it's rather `64 * 1024 * 1024 * 1024 / (64 * 1024) ` no?

Reply:
`var KB = 1024 * 1024;` :grimacing:

Reply:
I don't know, raising something from 256 to 1 million feels like the hell of a raise, that's why I felt weird about this

Reply:
Ah but this is just the second argument to `Math.ceil`, is this even a thing? I'm a bit lost there

Reply:
Correct me if I'm wrong; with `mem = 1024`:

```
Math.ceil(mem * 1024 / WASM_PAGE_SIZE) = 16;
```

therefore:
```
Math.min(Math.max(16, 1048576) + 256, 32767);
Math.min(1048832, 32767);
```

Due to the min/max logic, it seems like `totalMemory` (32767) will always win out regardless of `mem`.

Reply:
Sorry I had pending comments that I didn't confirm... (Not that used to github apparently x) ).
The min/max logic makes sure `initialMemory` doesn't go higher than 2GiB (`totalMemory`)  or lower than 256 + 256 pages.
My intentions were to bring up that lower value, but as @Alexandre-Herve pointed out, it could feel a bit weird to hard code it here; especially such a high value (64MiB). So instead I opted for adding a parameter to the algorithm so that the caller can choose to reserve enough the WASM memory at init so that subsequent calls would not require reallocation. Cf pull-request https://github.com/Dashlane/argon2-browser/pull/8.

I will close this one then. :)

Reply:
I think the initial upstream code is wrong. It should be :
```
var KiB = 1024;
var MiB = 1024 * KB;
var GiB = 1024 * MB;
var WASM_PAGE_SIZE = 64 * 1024;

 var totalMemory = (2 * GiB - 64 * KiB) / WASM_PAGE_SIZE;
var initialMemory = Math.min(Math.max(Math.ceil(params.mem * KiB/ WASM_PAGE_SIZE), 256) + 256, totalMemory);
```
It doesn't change anything to the outcome but it would better units-wise and would have avoided my 1024 factor issue here.